File: src/org/elixir_lang/ElixirParserDefinition.java
package org.elixir_lang;

import com.intellij.lang.ASTNode;
import com.intellij.lang.ParserDefinition;
import com.intellij.lang.PsiParser;
import com.intellij.lexer.Lexer;
import com.intellij.openapi.project.Project;
import com.intellij.psi.FileViewProvider;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.intellij.psi.tree.IFileElementType;
import com.intellij.psi.tree.TokenSet;
import org.elixir_lang.parser.ElixirParser;
import org.elixir_lang.psi.ElixirFile;
import org.elixir_lang.psi.ElixirTypes;
import org.jetbrains.annotations.NotNull;

/**
 * Created by luke.imhoff on 8/2/14.
 */
public class ElixirParserDefinition implements ParserDefinition {
    public static final TokenSet COMMENTS = TokenSet.create(ElixirTypes.COMMENT);
    public static final TokenSet STRING_LITERALS = TokenSet.create(
            ElixirTypes.CHAR_LIST_FRAGMENT,
            ElixirTypes.REGEX_FRAGMENT,
            ElixirTypes.SIGIL_FRAGMENT,
            ElixirTypes.STRING_FRAGMENT,
            ElixirTypes.WORDS_FRAGMENT
    );

    @NotNull
    @Override
    public Lexer createLexer(Project project) {
        return new ElixirLexer();
    }

    @NotNull
    public TokenSet getCommentTokens() {
        return COMMENTS;
    }

    @NotNull
    public TokenSet getStringLiteralElements() {
        return STRING_LITERALS;
    }

    @NotNull
    public PsiParser createParser(final Project project) {
        return new ElixirParser();
    }

    @Override
    public IFileElementType getFileNodeType() {
        return org.elixir_lang.psi.stub.type.File.INSTANCE;
    }

    public PsiFile createFile(FileViewProvider viewProvider) {
        return new ElixirFile(viewProvider);
    }

    public SpaceRequirements spaceExistanceTypeBetweenTokens(ASTNode left, ASTNode right) {
        SpaceRequirements spaceRequirements;

        if (left.getElementType() == ElixirTypes.KEYWORD_PAIR_COLON) {
            spaceRequirements = SpaceRequirements.MUST;
        } else {
            spaceRequirements = SpaceRequirements.MAY;
        }

        return spaceRequirements;
    }

    @NotNull
    public PsiElement createElement(ASTNode node) {
        return ElixirTypes.Factory.createElement(node);
    }
}

File: src/org/elixir_lang/ElixirParserDefinition.kt
package org.elixir_lang

import com.intellij.lang.ASTNode
import com.intellij.lang.ParserDefinition
import com.intellij.lang.PsiParser
import com.intellij.lexer.Lexer
import com.intellij.openapi.project.Project
import com.intellij.psi.FileViewProvider
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiFile
import com.intellij.psi.TokenType
import com.intellij.psi.tree.IFileElementType
import com.intellij.psi.tree.TokenSet
import org.elixir_lang.parser.ElixirParser
import org.elixir_lang.psi.ElixirFile
import org.elixir_lang.psi.ElixirTypes

class ElixirParserDefinition : ParserDefinition {
    override fun createElement(node: ASTNode): PsiElement = ElixirTypes.Factory.createElement(node)
    override fun createFile(viewProvider: FileViewProvider): PsiFile = ElixirFile(viewProvider)
    override fun createLexer(project: Project): Lexer = ElixirLexer(project)
    override fun createParser(project: Project): PsiParser = ElixirParser()
    override fun getCommentTokens(): TokenSet = COMMENTS
    override fun getFileNodeType(): IFileElementType = org.elixir_lang.psi.stub.type.File.INSTANCE
    override fun getStringLiteralElements(): TokenSet = STRING_LITERALS
    override fun getWhitespaceTokens(): TokenSet = WHITE_SPACES

    override fun spaceExistanceTypeBetweenTokens(left: ASTNode, right: ASTNode): ParserDefinition.SpaceRequirements =
            if (left.elementType === ElixirTypes.KEYWORD_PAIR_COLON) {
                ParserDefinition.SpaceRequirements.MUST
            } else {
                ParserDefinition.SpaceRequirements.MAY
            }

    companion object {
        val COMMENTS = TokenSet.create(ElixirTypes.COMMENT)
        val STRING_LITERALS = TokenSet.create(
                ElixirTypes.CHAR_LIST_FRAGMENT,
                ElixirTypes.REGEX_FRAGMENT,
                ElixirTypes.SIGIL_FRAGMENT,
                ElixirTypes.STRING_FRAGMENT,
                ElixirTypes.WORDS_FRAGMENT
        )
        val WHITE_SPACES = TokenSet.create(TokenType.WHITE_SPACE)
    }
}
